jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java

```diff
@@ -142,80 +143,120 @@ public void setLocalMaxDynamicTableSize(int localMaxDynamicTableSize)
         _localMaxDynamicTableSize = localMaxDynamicTableSize;
     }
 
-    public void encode(ByteBuffer buffer, MetaData metadata)
+    public boolean isValidateEncoding()
     {
-        if (LOG.isDebugEnabled())
-            LOG.debug(String.format("CtxTbl[%x] encoding", _context.hashCode()));
-
-        _headerListSize = 0;
-        int pos = buffer.position();
+        return _validateEncoding;
+    }
 
-        // Check the dynamic table sizes!
-        int maxDynamicTableSize = Math.min(_remoteMaxDynamicTableSize, _localMaxDynamicTableSize);
-        if (maxDynamicTableSize != _context.getMaxDynamicTableSize())
-            encodeMaxDynamicTableSize(buffer, maxDynamicTableSize);
+    public void setValidateEncoding(boolean validateEncoding)
+    {
+        _validateEncoding = validateEncoding;
+    }
 
-        // Add Request/response meta fields
-        if (metadata.isRequest())
-        {
-            MetaData.Request request = (MetaData.Request)metadata;
-
-            // TODO optimise these to avoid HttpField creation
-            String scheme = request.getURI().getScheme();
-            encode(buffer, new HttpField(HttpHeader.C_SCHEME, scheme == null ? HttpScheme.HTTP.asString() : scheme));
-            encode(buffer, new HttpField(HttpHeader.C_METHOD, request.getMethod()));
-            encode(buffer, new HttpField(HttpHeader.C_AUTHORITY, request.getURI().getAuthority()));
-            encode(buffer, new HttpField(HttpHeader.C_PATH, request.getURI().getPathQuery()));
-        }
-        else if (metadata.isResponse())
+    public void encode(ByteBuffer buffer, MetaData metadata) throws HpackException
+    {
+        try
         {
-            MetaData.Response response = (MetaData.Response)metadata;
-            int code = response.getStatus();
-            HttpField status = code < STATUSES.length ? STATUSES[code] : null;
-            if (status == null)
-                status = new HttpField.IntValueHttpField(HttpHeader.C_STATUS, code);
-            encode(buffer, status);
-        }
+            if (LOG.isDebugEnabled())
+                LOG.debug(String.format("CtxTbl[%x] encoding", _context.hashCode()));
 
-        // Remove fields as specified in RFC 7540, 8.1.2.2.
-        HttpFields fields = metadata.getFields();
-        if (fields != null)
-        {
-            // For example: Connection: Close, TE, Upgrade, Custom.
-            Set<String> hopHeaders = null;
-            for (String value : fields.getCSV(HttpHeader.CONNECTION, false))
+            HttpFields fields = metadata.getFields();
+            if (isValidateEncoding())
+            {
+                // Verify that we can encode without errors.
+                if (fields != null)
+                {
+                    for (HttpField field : fields)
+                    {
+                        String name = field.getName();
```

Comment:
strictly speaking we should not be checking the validity of ignored fields... not sure if the extra lookup costs more or checking anyway.